message.go

```diff
@@ -253,6 +254,47 @@ func DecodeMessageWithExt(b []byte, ext bool) (*Message, error) {
 			pos += int(extLen)
 		}
 	}
-	msg.Body = b[pos:]
+	if ext && msg.ExtVer == JSONHeaderExtVer {
+		json, err := msg.GetJsonExt()
+		if err != nil {
+			return nil, err
+		}
+		msg.Body, err = tryDecompress(b[pos:], json)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		msg.Body = b[pos:]
+	}
 	return &msg, nil
 }
+
+func tryDecompress(bodyMayCompressed []byte, ext *MsgExt) ([]byte, error) {
+	if ext.Custom[NSQ_CLIENT_COMPRESS_HEADER_KEY] != "" {
+		codecNo, err := strconv.Atoi(ext.Custom[NSQ_CLIENT_COMPRESS_HEADER_KEY])
+		if err != nil {
+			return nil, err
+		}
+		codec, err := GetNSQClientCompressCodeByCodecNo(codecNo)
+		if err != nil {
+			return nil, err
+		}
+		var originalMsgSize int
+		if originalMsgSizeStr, ok := ext.Custom[NSQ_CLIENT_COMPRESS_SIZE_HEADER_KEY]; ok {
+			originalMsgSize, err = strconv.Atoi(originalMsgSizeStr)
+			if err != nil {
+				return nil, err
+			}
+		} else {
+			//try luck with wild guess
+			originalMsgSize = len(bodyMayCompressed) * COMPRESS_RATIO_WILD_GUESS
```

Comment:
The buffer size may be too small, can we give a right size?

Reply:
It may. Logic here is a wild guess about original message size, once original message size could **not** be restored from message header. If everything works fine for producer compress, consumer won't hit this line with proper message size from header.